### PR TITLE
[FLINK-15636][python] Supports Python UDF in old planner under batch mode

### PR DESCRIPTION
**Pull request by hequn8128**
## What is the purpose of the change

Currently, Python UDF has been supported under old planner(only stream) and blink planner(stream&batch). This pull request dedicates to add Python UDF support in old planner under batch mode.

## Brief change log

  - Add rules to convert RelNode to DataSetPythonCalc Node.
  - Add PythonScalarFunctionFlatMap function to invoke Python ScalarFunctions for the legacy planner.
  - Add IT Tests.

## Verifying this change

This change added tests and can be verified as follows:

  - Added integration tests for Python UDF(test_udf.test_chaining_scalar_function) and dependency management(test_dependency.test_add_python_file).

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-python/src/main/java/org/apache/flink/table/runtime/functions/python/PythonScalarFunctionFlatMap.java [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+	 * Checks whether to invoke startBundle.
+	 */
+	private void checkInvokeStartBundle() throws Exception {
+		if (bundleStarted.compareAndSet(false, true)) {
+			pythonFunctionRunner.startBundle();
+		}
+	}
+
+	/**
+	 * Checks whether to invoke finishBundle by elements count. Called in processElement.
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/functions/python/PythonScalarFunctionFlatMap.java]: Correct the Java doc: Called in processElement -> Called in flatMap

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/CommonPythonCalc.scala [hunk 2]
```diff
@@ -89,4 +89,18 @@ trait CommonPythonCalc {
...
   }
+
+  private[flink] def getPythonRexCalls(calcProgram: RexProgram): Array[RexCall] = {
+    calcProgram.getProjectList
+      .map(calcProgram.expandLocalRef)
+      .collect { case call: RexCall => call }
+      .toArray
+  }
+
+  private [flink] def getForwardFields(calcProgram: RexProgram): Array[Int] = {
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/CommonPythonCalc.scala]: getForwardFields -> getForwardedFields

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/PythonCalcSplitRuleTest.scala [hunk 3]
```diff
@@ -266,4 +266,26 @@ class PythonCalcSplitRuleTest extends TableTestBase {
 
     util.verifyTable(resultTable, expected)
   }
+
+  @Test
+  def testSplitRuleForBatch(): Unit = {
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/PythonCalcSplitRuleTest.scala]: I guess we can remove this test case as the split rule is applied for both the streaming and batch plans. The original tests for streaming plans are enough. What do you think?

**Comment by hequn8128** [comment]
@dianfu Thanks a lot for the review and suggestions. I have addressed your comments and updated the PR.

**Comment by dianfu** [comment]
@hequn8128  Thanks for the update. LGTM. Merging...

**Pull request closed without merging**